### ISSUE TITLE
Asynchronous events (take 3)

### ISSUE THREAD

**Issue by SOF3** [pmmp/PocketMine-MP#2470]
<details open>
<summary>TL;DR</summary>

Call an async event like this:

```php
/** @var \pocketmine\plugin\PluginManager $pluginManager */
/** @var \pocketmine\event\Event $event */
$pluginManager->callAsyncEvent($event, function($event) {
  // executed after all async handlers have completed
  // $event is identical to the $event passed, just for simplicity
});
```

Handle an async event like this:

```php
use pocketmine\event\{AsyncEvent, Listener};

class SomeListenerClass implements Listener {
  // $event is the dispatched event
  // $async manages the async semantics. There is a specific instance for each handler.
  public function arbitrary_method_name(Event_Class_Name $event, AsyncHandler $async) : void{
    // to execute something asynchronous:
    $async->block(); // stop calling any other handlers until $async->unblock() is called
    // alternatively:
    $async->block(EventPriority::HIGH); // all handlers at EventPriority::HIGH or higher will not be called until $async->unblock() is called
    // note that recursive blocking is not allowed

    some_operation_that_callbacks_later(function($result) use($event, $async, $id) : void{
      $async->unblock();
      // unblock() will not immediately dispatch the handlers afterwards. It is still safe to modify the event here.
      $event->setResult($result);
    });
  }
}
```

</details>

# Full RFC
## Definition of terms
- Requirement levels follow [RFC 2119](https://tools.ietf.org/search/rfc2119) and [RFC 6919](https://tools.ietf.org/html/rfc6919) when used in UPPERCASE.
- "listener" vs "handler": "listener" refers to the _class_ that implements pocketmine\event\Listener, while "handler" refers to the _methods_ in the listener that handle events.
- "async": it refers to the generic concept in software programming that "some function created previously is called after another operation has completed". It has no specific relation to threading or AsyncTask.
- "immediate" is the antonym of "async", implying that something takes place within the function call.

## Motivation
<details open>
<summary>Why async events?</summary>

It is difficult to write a plugin that fetches data asynchronously because PocketMine was designed in a direct approach such that all operations take place immediately. In particular, the events API expects the event data to be modified immediately. Plugins usually have to assume the result with less side effects and correct the side effects when the async result arrives. Take these two examples:

For example, a plugin checks whether a player is banned in PlayerPreLoginEvent, _asynchronously_. The plugin assumes the player has not been banned (because otherwise kicking the player is an irreversible action), i.e. PlayerPreLoginEvent is not cancelled, and the player logins as usual, possibly even joining the server. After the plugin fetches the data and confirms that the player should be banned, the player is kicked late.

Another example is that a plugin checks a chat message online in PlayerChatEvent, _asynchronously_. The plugin assumes the chat message is invalid (because otherwise approving and sending the message is an irreversible action), i.e. PlayerChatEvent is cancelled.  After the plugin fetches the data and confirms that the message is valid, it sends the message again.

There are two problems with this approach:
- Effect on other handlers: An assumed value may incorrectly induce the behaviour of other handlers of the same event
- Effect on semantics: The assumed value also has its side effects, e.g. in the first example, the player can still join the server and potentially cause destruction, or simply wasting resources sending data.

This RFC seeks to provide a better solution by providing specialized async support from the events API.

</details>

## Relevant issues
- This replaces and obsoletes #1881

<details open>
<summary>How is this different from #1881?</summary>

- This RFC adds the ability to execute multiple event handlers in parallel.
- This RFC separates async semantics from the event class to enhance backwards compatibility and [reduces inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).
- This RFC reduces dependency on the Server, enhancing the possibility of unit testing.

</details>

### Phase I: The async event API
<details open>

#### Allow an event to be called asynchronously
Events that may be called asynchronously should implement the interface `AsyncEvent`.

#### Call an event asynchronously
Calling an event asynchronously implies that the event handlers will not finish completing when the callAsyncEvent() method returns. Instead, a callable is provided and is called when the last handler completes:

```php
$pluginManager->callAsyncEvent($event, $callback);
```

To facilitate code fluency, `$callback` may accept an `Event` instance, which is identical to the `$event` instance passed in.

#### Non-async handler for async event
Async events can be handled in a non-async manner to allow backwards compatibility. The syntax and semantics are identical to non-async events.

#### Async handler for async event
An async handler should declare two arguments: `$event` and `$async`. `$event` is the argument as in non-async handlers, while `$async` must be type-hinted with `AsyncHandler`, which is an object responsible for the async behaviour of each handler.

Async handlers can call `$async->block()` (`$async->block(null)`). After the handler returns, until `$async->unblock()` is called, no new handlers will be invoked for the same event. 

In addition, async handlers can call `$async->block($priority)`, where `$priority` is an int constant from the `EventPriority` class. Until `$async->unblock()` is called, no handlers of priority at or higher than `$priority` is invoked. Other event handlers are invoked as usual. This allows parallel execution of multiple async handlers, but might lead to undefined behaviour with event states. Therefore, async handlers MUST always check if the handler state is still valid at the point of unblocking, no matter it is priority-blocking or null-blocking.

#### Blocking timeout
The `AsyncEvent` interface may contain a `getMaxTimeout()` method, which specifies the maximum time allowed per handler (in seconds). This should return 5.0 by default.

If the timeout exceeds, the handler would be disabled, a warning would be logged and the handler is assumed complete. Further calls to `unblock()` will do nothing.

#### Multiple event calls
Each event instance may only be called again after the all handlers have completed.

</details>

### Phase II: Call normal events asynchronously
The second phase is to call events from the PocketMine API using the async-events API. Whether each event should be async will be discussed at a later stage due to its complexity beyond the scope of phase I.

## Assessments
(Unfinished)

### Performance
<details open>

</details>

### Backwards compatibility
<details open>

</details>

## Proposed API version
The addition of phase I is proposed to be added in minor.next as it contains no significant backwards incompatibility.

## Expected use cases
The async-events RFC is an abstract model for event dispatching. The details of each async event, such as time allowed, should be defined by the event callers. However, in general, handlers MUST NOT block events indefinitely. A time limit of a few seconds or a few minutes are reasonable depending on the scenario. These are some general suggestions:
- If an event is derived from a very frequent incident on the same subject (&ge;1 Hz per player), callers SHOULD NOT call the event asynchronously at all to prevent stacking multiple handler calls concurrently.
- If an event is derived from a frequent incident (&ge;0.1 Hz), handlers SHOULD NOT block the event for more than the time for one or two HTTP roundtrips.
- If an event is derived from explicit user input (e.g. commands), or from very rare incidents (once per &ge;10 minutes), handlers MAY block the event _definitely_ to wait for simple user input. If the user does not respond, the event should be unblocked within a few minutes.

Handlers MUST NOT abuse async events into a time scheduler. For example, it is very abusive to create a chat throttle by blocking a PlayerChatEvent and unblocking it a few seconds later. This is because it will result in an indefinite stack of async events that could have negative impact on performance.

Handlers MUST also NOT abuse async events into a delayed logic controller. For example, it is not reasonable to handle a PlayerLevelChangeEvent by first placing the player in a queue of deathmatch arena to qualify for switching levels.

**Comment by SOF3** [comment]
I forgot to clarify:

Async is, therefore, determined by 3 parts: declaration `implements AsyncEvent`, calling `callAsyncEvent` and handling `AsyncHandler $async`. This table describes the behaviour:

| implements `AsyncEvent` | `callAsyncEvent` | has `AsyncHandler $async` | behaviour |
| :---: | :---: | :---: | :---: |
| F | F | F | behaves as usual non-async events |
| F | F | T | error while `registerEvents`: "$class does not implement AsyncEvent and cannot be handled asynchronously |
| F | T | F | error while `callAsyncEvent`: "$class does not implement AsyncEvent and cannot be handled asynchronously" |
| T | F | F | behaves as usual non-async events |
| T | F | T | the handler is not invoked, but no errors are triggered; a DEBUG-level notice may be triggered |
| T | T | F | the handler behaves as usual non-async events |
| T | T | T | the event is called asynchronously, allowing the use of block() and unblock() |

**Comment by SOF3** [comment]
Another clarification: `block(priority)` is equivalent to not blocking the current handler, then insert a handler that blocks until unblock() is called to the end of the handler list at `priority`.

The initial RFC was therefore written wrongly. `block(EventPriority::HIGH)` will not block handlers at HIGH, but only handlers at HIGHEST and MONITOR.

If the handler is blocked at EventPriority::MONITOR, it shall not result in any

**Comment by SOF3** [comment]
I am proposing the following solution to the question above:
```php
	/**
	 * Instructs the event dispatch chain to stop if the next event handler is higher than the specified priority until
	 * unblock() has been called on this object.
	 * (Event handlers <em>exactly at</em> $maxPriority will still be executed)
	 *
	 * If $maxPriority is null, the event dispatch chain will stop executing <em>any</em> new handlers until unblock()
	 * has been called. This is similar to passing a priority lower than LOWEST.
	 *
	 * To summarize,
	 * - passing `null` will always block all new handlers from being invoked.
	 * - passing the current handler's priority will still finish handlers at the current priority
	 * - passing `EventPriority::MONITOR` will not block any event handlers, but the event caller will not receive the
	 *   event until this event has been unblocked.
	 */
```
